contracts/Crowdfunding.sol

```diff
@@ -5,99 +5,220 @@ pragma solidity ^0.8.0;
+        string memory _imageURL
+    ) external {
+        require(bytes(_title).length > 0, "Title cannot be empty");
+        require(bytes(_description).length > 0, "Description cannot be empty");
+        require(_deadline > block.timestamp, "Deadline must be in the future");
+        require(_goal > 0, "Funding goal cannot be zero");
+
+        Campaign memory campaign;
+        campaign.id = numCampaigns;
+        numCampaigns += 1;
+        campaign.creator = msg.sender;
+        campaign.title = _title;
+        campaign.description = _description;
+        campaign.goal = _goal;
+        campaign.deadline = _deadline;
+        campaign.imageURL = bytes(_imageURL).length > 0
+            ? _imageURL
+            : "URL of a default placeholder";
+        campaign.raisedAmount = 0;
+        campaign.state = State.OPEN;
+        campaign.contributors = new address[](0);
+
+        campaigns.push(campaign);
+
+        emit CampaignCreated(numCampaigns, msg.sender, _goal, _deadline);
+    }
 
-        return campaigns[_campaignId];
+    // Function to contribute funds to a campaign
+    // the user pay ethers to the contract when calling this function (msg.value)
+    function contribute(uint256 _campaignId) external payable {
+        require(
+            msg.sender != campaigns[_campaignId].creator,
+            "Creator can not contribute to its own campaign"
+        );
+        require(
+            campaigns[_campaignId].state == State.OPEN,
+            "Campaign is not open"
+        );
+        require(
+            block.timestamp < campaigns[_campaignId].deadline,
+            "Campaign deadline has passed"
+        );
+        /* require(
+            campaigns[_campaignId].raisedAmount + _amount <=
+                campaigns[_campaignId].goal,
+            "Contribution exceeds campaign goal"
+        ); cam raise more money than goal*/
+
+        campaigns[_campaignId].raisedAmount += msg.value;
+        user2Campaigns[msg.sender].push(_campaignId);
+        campaigns[_campaignId].contributors.push(msg.sender);
     }
 
-    function getCampaigns() public view returns(Campaign[] memory) {
-        Campaign[] memory allCampaigns = new Campaign[](numCampaigns);
+    // Function to close a campaign
+    function closeCampaign(
+        uint256 _campaignId
+    ) external onlyCreator(_campaignId) {
+        require(
+            campaigns[_campaignId].state == State.OPEN,
+            "Campaign is not open"
+        );
+        require(
+            block.timestamp < campaigns[_campaignId].deadline,
+            "Campaign deadline has passed"
+        );
 
-        for (uint256 i = 0; i < numCampaigns; i++) {
-            allCampaigns[i] = campaigns[i];
+        // Check if the goal is met
+        if (
+            campaigns[_campaignId].raisedAmount >= campaigns[_campaignId].goal
+        ) {
+            // Release funds to campaign creator
+            withdraw(_campaignId);
+        } else {
+            refund(_campaignId);
```

Comment:
really makes more sense to be campaign.refund(), but this is how our structure is lol

Reply:
yes. we should create another contract for campaign lol

Reply:
yes. we should create another contract for campaign lol